cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -259,223 +252,88 @@ type PaginationOptions struct {
+		ledgerCloseTimestamp int64
+		event                xdr.ContractEvent
+	}
+	var found []entry
+	err := scanner.Scan(
+		events.Range{
+			Start:      start,
+			ClampStart: false,
+			End:        events.MaxCursor,
+			ClampEnd:   true,
+		},
+		func(cursor events.Cursor, ledgerCloseTimestamp int64, event xdr.ContractEvent) bool {
+			if request.Matches(event) {
+				found = append(found, entry{cursor, ledgerCloseTimestamp, event})
 			}
-		}
-		return nil
-	})
-	if err == io.EOF {
-		err = nil
+			return len(found) < limit
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not scan events")
 	}
-	return results, err
-}
 
-// ForEachTransaction runs f for each transaction in a range from start
-// (inclusive) to finish (exclusive). If f returns any error,
-// ForEachTransaction stops immediately and returns that error.
-func (a EventStore) ForEachTransaction(start, finish *toid.ID, f func(transaction horizon.Transaction) error) error {
-	delay := 10 * time.Millisecond
-	cursor := toid.New(start.LedgerSequence, start.TransactionOrder, 0)
-	for {
-		transactions, err := a.Client.Transactions(horizonclient.TransactionRequest{
-			Order:         horizonclient.Order("asc"),
-			Cursor:        cursor.String(),
-			Limit:         200,
-			IncludeFailed: false,
-		})
+	var results []EventInfo
+	for _, entry := range found {
+		info, err := eventInfoForEvent(
+			entry.cursor,
+			time.Unix(entry.ledgerCloseTimestamp, 0).UTC().Format(time.RFC3339),
+			entry.event,
+		)
 		if err != nil {
-			hErr := horizonclient.GetError(err)
-			if hErr != nil && hErr.Response != nil && (hErr.Response.StatusCode == http.StatusTooManyRequests || hErr.Response.StatusCode >= 500) {
-				// rate-limited, or horizon server-side error, we can retry.
-
-				// exponential backoff, to not hammer Horizon
-				delay *= 2
-
-				if delay > time.Second {
-					return err
-				}
-
-				// retry
-				time.Sleep(delay)
-				continue
-			} else {
-				// Unknown error, bail.
-				return err
-			}
-		}
-
-		for _, transaction := range transactions.Embedded.Records {
-			pt, err := strconv.ParseInt(transaction.PagingToken(), 10, 64) //lint:ignore gomnd
-			if err != nil {
-				return errors.Wrapf(err, "invalid paging token %s", transaction.PagingToken())
-			}
-			if pt >= finish.ToInt64() {
-				// Done!
-				return nil
-			}
-			id := toid.Parse(pt)
-			cursor = &id
-
-			if err := f(transaction); err != nil {
-				return err
-			}
-		}
-
-		if len(transactions.Embedded.Records) < 200 {
-			// Did not return "limit" transactions, and the query is open-ended, so this must be the end.
-			return nil
+			return nil, errors.Wrap(err, "could not parse event")
 		}
+		results = append(results, info)
 	}
+	return results, nil
 }
 
-func eventInfoForEvent(opID *toid.ID, eventIndex int32, ledgerClosedAt string, event xdr.ContractEvent) (EventInfo, error) {
+func eventInfoForEvent(cursor events.Cursor, ledgerClosedAt string, event xdr.ContractEvent) (EventInfo, error) {
```

Comment:
could you reorder the arguments 
event xdr.ContractEvent, cursor events.Cursor, ledgerClosedAt string
?
right now, the first two are only being used when constructing the output EventInfo.
given that the entire function is about the "event", it should be the first argument.